### ISSUE TITLE
AMQP-WS device disconnect issues. 

### ISSUE THREAD

**Issue by mikevax** [Azure/azure-iot-sdk-node#70]
We are using amqp-ws. In A streaming job in Azure we are monitoring the Operations monitoring endpoint. What we see are disconnects about every 45 min. however we do not see disconnect event in our nodeJS application running on our IoT device. Are there any update coming for the amqp-ws nodejs library. Last updated September 2016.

We are using  Operations monitoring endpoint to determine if out device is online. periodically, not always we get a connect followed by a disconnect. Sometimes with in the same 10 seconds and dod not get a nodejs disconnect event.
 

Linux/Yocto
NodeJS 6.11.0
NPM version 3.10.10

List of Pachages
Relevant:
    "azure-iot-device": "^1.1.16",
    "azure-iot-device-amqp": "^1.1.16",
    "azure-iot-device-amqp-ws": "^1.0.16",
  

All
{
  "name": "mavo-torii",
  "version": "8.5.0",
  "accVersion": "6.2.0.1",
  "description": "Mavo-Torii - ACC's Web Facade",
  "main": "./index.js",
  "scripts": {
    "start": "gulp serve",
    "production": "node --max-old-space-size=250 --abort-on-uncaught-exception ./watchdog.js",
    "changelog": "gulp changelog",
    "test": "gulp test",
    "lint": "gulp lint",
    "gulp": "gulp",
    "testApi": "node test/testApi --transport=rest --domains=alarms --address=10.40.237.2",
    "artillery": "artillery quick --duration 60 --rate 10 -n 20 http://localhost:8080/api/v1/PushService/Notifications"
  },
  "author": "Avigilon Corporation",
  "contributors": [
    {
      "name": "Dan BROOKS",
      "email": "dan.brooks@avigilon.com",
      "url": "http://twitter.com/danactive"
    },
    {
      "name": "Tomer Goldenberg",
      "email": "tomer.goldenberg@avigilon.com"
    }
  ],
  "license": "Avigilon Corporation",
  "engines": {
    "node": "6.10.3",
    "npm": ">=3.10.10"
  },
  "private": true,
  "devDependencies": {
    "archiver": "^1.3.0",
    "boom": "^4.3.1",
    "deep-equal": "^1.0.1",
    "del": "^2.2.2",
    "eslint": "^3.19.0",
    "eslint-config-airbnb-base": "^11.1.2",
    "eslint-plugin-header": "^1.0.0",
    "eslint-plugin-import": "^2.2.0",
    "exceljs": "^0.4.2",
    "faux-jax": "^5.0.6",
    "file": "^0.2.2",
    "file-exists": "^3.0.2",
    "filter-object": "^2.1.0",
    "gulp": "^3.9.1",
    "gulp-cli": "^1.2.2",
    "gulp-count": "^0.3.0",
    "gulp-eslint": "^3.0.1",
    "gulp-filter": "^5.0.0",
    "gulp-header-license": "^1.0.9",
    "gulp-if": "^2.0.2",
    "gulp-install": "^1.1.0",
    "gulp-istanbul": "^1.1.1",
    "gulp-jsdoc3": "^1.0.1",
    "gulp-newer": "^1.3.0",
    "gulp-nsp": "^2.4.2",
    "gulp-rename": "^1.2.2",
    "gulp-strip-comments": "^2.4.5",
    "gulp-tape": "0.0.9",
    "gulp-util": "^3.0.8",
    "inquirer": "^3.0.6",
    "istanbul": "^0.4.5",
    "jpeg-js": "^0.2.0",
    "license-extractor": "^1.0.4",
    "lodash.memoize": "^4.1.2",
    "long": "^3.2.0",
    "merge-stream": "^1.0.1",
    "minimist": "^1.2.0",
    "readable-stream": "^2.2.6",
    "run-sequence": "^1.2.2",
    "semver": "^5.3.0",
    "server-destroy": "^1.0.1",
    "shankwrap": "^2.0.0",
    "shelljs": "^0.7.7",
    "sinon": "^1.17.6",
    "socket.io-client": "^1.7.3",
    "ssh2": "^0.5.4",
    "tap-parser": "^5.3.3",
    "tape": "^4.6.3",
    "tape-catch": "^1.0.6",
    "through2": "^2.0.3",
    "toxy": "^0.3.12",
    "which": "^1.2.14",
    "xmlbuilder": "^8.2.2"
  },
  "dependencies": {
    "async": "^2.3.0",
    "azure-iot-device": "^1.1.16",
    "azure-iot-device-amqp": "^1.1.16",
    "azure-iot-device-amqp-ws": "^1.0.16",
    "blankie": "^2.0.0",
    "bonjour": "https://github.com/ionic-team/bonjour.git#518c00cc7ff0644039d96b8b07d0ca95e719194e",
    "bytebuffer": "^5.0.1",
    "clone": "^2.1.1",
    "conf": "^0.12.0",
    "dateformat": "^2.0.0",
    "deep-freeze": "0.0.1",
    "deepmerge": "^1.3.2",
    "dot-prop": "^4.1.1",
    "dustjs-linkedin": "~2.7.5",
    "end-of-stream": "^1.4.0",
    "file-exists": "^3.0.2",
    "form-data": "^2.1.4",
    "fs-extra": "^2.1.2",
    "glob": "^7.1.1",
    "good": "^7.1.0",
    "good-winston": "^3.1.0",
    "hapi": "^16.1.1",
    "hapi-and-healthy": "^4.6.1",
    "hapi-auth-basic": "^4.2.0",
    "hapi-dust": "^2.0.0",
    "hapi-swagger": "^7.7.0",
    "http-auth": "^3.1.3",
    "http-status-codes": "^1.1.6",
    "inert": "^4.2.0",
    "joi": "^10.4.1",
    "js-yaml": "^3.8.2",
    "lout": "^10.0.1",
    "minimist": "^1.2.0",
    "moment": "^2.18.1",
    "node-forge": "^0.7.1",
    "node-netstat": "^1.4.1",
    "node-ssdp": "^3.2.1",
    "node-uuid": "^1.4.8",
    "protobufjs": "^6.7.0",
    "request": "^2.81.0",
    "retry": "^0.10.0",
    "scooter": "^4.0.0",
    "socket.io": "^1.7.3",
    "stream-buffers": "^3.0.1",
    "string-template": "^1.0.0",
    "vision": "^4.1.1",
    "winston": "^2.3.1",
    "wreck": "^12.0.0",
    "xml2js": "^0.4.17",
    "xml2js-xpath": "^0.8.0"
  },
  "optionalDependencies": {
    "memwatch-next": "^0.3.0"
  }
}

# Context
- **OS and version used:** <VERSION> 
- **Node.js version:** <VERSION> 
- **npm version:** <VERSION> 
- **list of installed packages:** <VERSION> 
- **cloned repo:** <VERSION> 

# Description of the issue:

# Code sample exhibiting the issue:

# Console log of the issue:

**Comment by pierreca** [comment]
Let me rephrase what I think I understand:
- you are using some event hubs SDK to monitor the operations-monitoring endpoint and see devices disconnecting every 45 minutes
- on the device side you're using the AMQP/WS transport
- the ops monitoring endpoint shows specifically those devices (running over amqp/ws) disconnecting every 45 minutes
- but on the device side, you're not seeing the device being disconnected

Additional questions:
- What type of authentication are you using on the device?
- Are you regularly sending or receiving messages/do you see them coming in on the event hubs compatible endpoint?
- how do you instantiate the client object on the device (code snippet pls?)

**Comment by pierreca** [comment]
@mikevax did you get through this? i asked a couple questions here but didn't here from you?

**Comment by mikevax** [comment]
Hi, back on this issues:
  - What type of authentication are you using on the device?
     We make an out of band https:// request to our cloud service that returns the IoT hub keys for the device. The authentication used is SAS keys maintained by the iot hub  that expire every 45 min.

   - Are you regularly sending or receiving messages/do you see them coming in on the event hubs compatible endpoint?
      The short answer is yes, however sometimes the device may not be sending events for hours. Currently we are using a streaming job with the input of the iot hub metrics. We sometimes  see a disconnect after a SAS key update without a connect message. We are using this information to determine if the device  is online. A work around for this issue it to query the iot metrics to see if the device has been down for 5 min. If this is the case se send a ping to the device and the device responds with a pong.  In no pong response occurs within 10 second or so we declare the device down. We were hoping to use this mechanism to replace a costly heartbeat message. **Bottom line, we would like to rely on the iot hub metrics to determine if a device is down.** One solution is to eliminate the SAS key refresh by using certificates, however this is a future development task after our fist release.

- how do you instantiate the client object on the device (code snippet pls?)
Below is a simplified version in Node of how we create the client minus all the options and error handling.
const ProtocolAmqpWs = require('azure-iot-device-amqp-ws');
 clientFromConnectionString = ProtocolAmqpWs.clientFromConnectionString;
 // the connection string is created in the Azure cloud after creating an iot device and send to the client during the provisioning step of the device
  var client = clientFromConnectionString(connectionString);

**Comment by pierreca** [comment]
@mikevax thanks for the details
based on my understanding of what you see, here's what I think is happening: you're using an older version of the SDK, that is using SASL-PLAIN authentication to renew the SAS token created automatically by the SDK from the connection string that the device gets from your cloud service.

A couple elements have changed since then if you update:
- we now use CBS for authentication over AMQP and AMQP/WS: therefore you won't see any disconnects anymore if you update, except if the device connection fails
- Like you said another solution is to use an x509 cert which doesn't require maintaining CBS links.
- Yet another solution (that would even probably be slightly more secure) would be to have your service keep the secret keys secret, and instead distribute shared access tokens with an expiry time that you control, and then instantiate your client with `Client.fromSharedAccessSignature()` instead of `fromConnectionString()`. (more secure because your secrets aren't sent over a wire, only temporary tokens).

In any case, this disconnect/reconnect thing should be a thing of the past if you update - please let us know if that fixes the issue.

**Comment by mikevax** [comment]
Thanks for the quick response:

    When you refer to SDK are you referring to the Azure SDK or node client SDK or both ?
Exactly what version should I be using? for Azure and node.
     Also if you have examples of using CBS that would be great.
For node we are using:
azure-iot-device": "^1.1.16",
"azure-iot-device-amqp": "^1.1.16",
"azure-iot-device-amqp-ws": "^1.0.16",

For Azure sdk we use:
1.0.1.15

**Comment by mikevax** [comment]
Follow up question:

   In the case where we use Client.fromSharedAccessSignature() what secure method would you use to renew the SAS signature when the device has been offline and the key expired. If we expose an API to renew the key given the device id it potentially opens a security hole.

**Comment by pierreca** [comment]
For the API for shared access signature, it's technically the same API that you would use for the symmetric key, isn't it? just give it a shared access signature instead of a symmetric key? Before we explore this though I feel there's something I don't get about what you're observing:

The CBS code is automatically active, no need to do anything on your side. we introduced it in [version 1.1.8](https://github.com/Azure/azure-iot-sdk-node/releases/tag/2017-3-24). so you should've been good to go and there shouldn't be any disconnection every 45 minutes.

What you can see [here](https://github.com/Azure/azure-iot-sdk-node/blob/master/device/transport/amqp/src/amqp.ts#L355) is that we set the `needToReconnect` parameter of the `SharedAccessSignatureUpdated` result to `false` which means the client will not actively try to disconnect and reconnect the transport [here](https://github.com/Azure/azure-iot-sdk-node/blob/master/device/core/src/client.ts#L300).

So if you're still observing a disconnect/reconnect, then we have a bug somewhere. either in the SDK, or in the Operations Monitoring code, but somewhere, something still isn't acting or being reported right.

Would it be possible for you to activate more debugging traces on the device side? This is done by setting an environment variable called `DEBUG`. you can set it to `device-amqp:amqp` at first and if this doesn't help us narrow it down we can set it to `*`. Be careful before posting any logs here though, since there are amqp frames in there it might contain credentials (or just use a temporary key for the device you're logging from)

**Comment by mikevax** [comment]
Thanks for you response, you have been very helpful.
A few additional questions/comments:
1.looking at the latest official releases for node we should use the release 2017-10-09 correct:
versions contained in the below link
https://github.com/Azure/azure-iot-sdk-node/commit/20c7f20e8682a06f58eeaf36f11ee72d5c5c8415
2. for the c# sdk should we use Microsoft.Azure.Amqp --version 2.1.2 ?
3. could you explain what we need to do on the sever and device side to use secret key/temporary tokens ?
4. exactly what make secret keys more secure for the case of device impersonation ?
you previous comment:
Yet another solution (that would even probably be slightly more secure) would be to have your service keep the secret keys secret, and instead distribute shared access tokens with an expiry time that you control, and then instantiate your client with Client.fromSharedAccessSignature() instead of fromConnectionString(). (more secure because your secrets aren't sent over a wire, only temporary tokens).

**Comment by pierreca** [comment]
> 1.looking at the latest official releases for node we should use the release 2017-10-09 correct:
versions contained in the below link 20c7f20

yes.

> 2. for the c# sdk should we use Microsoft.Azure.Amqp --version 2.1.2 ?

I don't know - This sounds like the Amqp library itself, I'd rather refer you to the [C# SDK github repo](https://github.com/Azure/azure-iot-sdk-csharp/releases): 

> 3. could you explain what we need to do on the server and device side to use secret key/temporary tokens ?

On the client side: use `Client.fromSharedAccessSignature()` instead of `Client.fromConnectionString()`. On the server side: generate shared access signatures instead of passing the symmetric key. (see [this dev guide](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security)

> 4. exactly what make secret keys more secure for the case of device impersonation ?
you previous comment:
Yet another solution (that would even probably be slightly more secure) would be to have your service keep the secret keys secret, and instead distribute shared access tokens with an expiry time that you control, and then instantiate your client with Client.fromSharedAccessSignature() instead of fromConnectionString(). (more secure because your secrets aren't sent over a wire, only temporary tokens).

A shared access signature token is temporary. after expiry time, anyone who got their hands on it can't use it. a SymmetricKey doesn't have an expiry date and therefore must be rotated if compromised. (see previous link)

**Comment by mikevax** [comment]
Good afternoon,

    Per your suggestion we update to the latest server and device components for the device and server SDK. We still see the same device drops every 45 min. How can we debug the issues from here?
Note we are using AMQP-WS web sockets over port 443.

Thanks,

Again our code snippet to create the connection.
const ProtocolAmqpWs = require('azure-iot-device-amqp-ws');
clientFromConnectionString = ProtocolAmqpWs.clientFromConnectionString;
// the connection string is created in the Azure cloud after creating an iot device and send to the client during the provisioning step of the device
var client = clientFromConnectionString(connectionString);

Device Versions from the October 9, 2017.
azure-iot-device package file
{
  "raw": "azure-iot-device@^1.1.19",
  "scope": null,
  "escapedName": "azure-iot-device",
  "name": "azure-iot-device",
  "rawSpec": "^1.1.19",
  "spec": ">=1.1.19 <2.0.0",
  "type": "range"
},
Our node app package file:
    "azure-iot-device": "^1.1.19",
    "azure-iot-device-amqp": "^1.2.0",
    "azure-iot-device-amqp-ws": "^1.0.16",

On the Azure side we are using the 1.4 devices Azure iot-device sdk

**Comment by pierreca** [comment]
@mikevax we finally shipped a fully-fledged retry logic for the device SDK
- https://github.com/Azure/azure-iot-sdk-node/releases/tag/2017-10-24
- https://www.npmjs.com/package/azure-iot-device (`azure-iot-device@1.2.0`)

Apologies for the code churn, this is the latest release of a series that completely refactored the SDK to be more stable, reconnect automatically and retry pending operations.

**Comment by mikevax** [comment]
Good Morning,

    We have tested using azure-iot-device@1.2.0 version and we are having the same issues with the disconnects.
azure-iot-device package file used:
"_args": [
    [
      {
        "raw": "azure-iot-device@1.2.0",
        "scope": null,
        "escapedName": "azure-iot-device",
        "name": "azure-iot-device",
        "rawSpec": "1.2.0",
        "spec": "1.2.0",
        "type": "version"
      },
      "C:\\Helious\\master\\mavo-torii"
    ]
  ],
  "_from": "azure-iot-device@1.2.0",
  "_id": "azure-iot-device@1.2.0",
  "_inCache": true,
  "_location": "/azure-iot-device",
  "_nodeVersion": "6.11.0",
  "_npmOperationalInternal": {
    "host": "s3://npm-registry-packages",
    "tmp": "tmp/azure-iot-device-1.2.0.tgz_1508898136391_0.8665243561845273"
  },

How do we proceed from here it temes of resolving issues?  Are there breakpoints we can set in the iot-device code to dtermine what the issues are ?
 We are getting close to shipping and need to resolve iot disconnect issues this week. Should we open a Microsoft ticket for this issues?
  If you think a conversation or a debug session would help you can reach me at 617-877-3332.
Thanks,
Mike

**Comment by pierreca** [comment]
@mikevax 
There's something you mentionned earlier that I didn't pick up: 
you say you're using `azure-iot-device-amqp-ws@1.0.16` - this package was actually deprecated some time ago and is replaced by the AmqpWs object in  `azure-iot-device-amqp`. You can simply replace your `require` calls this way (and no other code change should be necessary):

```
// Replace:
var AmqpWs = require('azure-iot-device-amqp-ws').AmqpWs;

// With:
var AmqpWs = require('azure-iot-device-amqp').AmqpWs
```

I've just set up a hub with the latest packages and a CBS token renewal interval of 10 seconds and I see now disconnections whatsoever on the ops monitoring endpoint. 

I thought we had removed all references to this package a while ago and marked is as deprecated. I'll double check the latter, but can you let me know if you found a reference to it somewhere?

**Comment by mikevax** [comment]
Good news, your suggestion to remove the reference to azure-iot-amqp-ws and replace with azure-iot-amqp worked!

Also there should be a note on the npm page https://www.npmjs.com/package/azure-iot-device-amqp-ws the library is deprecated.

Thanks for the help, glad to finally resolve this issue.

Followup question: Back to using SAS keys vs connection strings. Exactly what needs to be implemented to refresh the SAS keys on the Azure cloud side and the client side. Thanks,

**Comment by pierreca** [comment]
Glad the issue is solved. I'm closing it.

Regarding the SAS token vs Symmetric Key discussion: that'd deserve a separate thread but I can get you started here. The bottom line is: 
- SAS tokens are temporary and expire
- Symmetric Keys are forever and have to rolled in the device registry if they are compromised.

which is why if you have an API that a device can call to get its credentials, then I'd rather suggest having this API returns SAS tokens than Symmetric Keys. The problem is your devices are going to hit this API a lot, depending on how often you want to renew the SAS tokens (you have to renew them since they expire, unlike the symmetric key).

If you want to generate SAS tokens, the SDK can help you do that using the `SharedAccessSignature`  object. The code of the iothub-explorer tool has a good example of how to do this: https://github.com/Azure/iothub-explorer/blob/master/iothub-explorer-sas-token.js

if you want to read more about the differences between SAS tokens and symmetric keys, I would recommend reading this:
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security

Sorry I didn't notice the package thing earlier. I've marked it as deprecated, and I'll see if we can publish a new version with a readme explicitely stating that it is deprecated.
